dali/python/nvidia/dali/plugin/base_iterator.py

```diff
@@ -0,0 +1,185 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from nvidia.dali import types
+import math
+import logging
+import numpy as np
+
+class _DaliBaseIterator(object):
+    def __init__(self,
+                 pipelines,
+                 size=-1,
+                 reader_name=None,
+                 auto_reset=False,
+                 fill_last_batch=True,
+                 last_batch_padded=False):
+
+        assert pipelines is not None, "Number of provided pipelines has to be at least 1"
+        if not isinstance(pipelines, list):
+            pipelines = [pipelines]
+        self._num_gpus = len(pipelines)
+        # frameworks expect from its data iterators to have batch_size field,
+        # so it is not possible to use _batch_size instead
+        self.batch_size = pipelines[0].batch_size
+        self._size = int(size)
+        self._auto_reset = auto_reset
+
+        self._fill_last_batch = fill_last_batch
+        self._last_batch_padded = last_batch_padded
+        assert self._size != 0, "Size cannot be 0"
+        assert self._size > 0 or (self._size < 0 and (len(pipelines) == 1 or reader_name)), "Negative size is supported only for a single pipeline"
+        assert not reader_name or (reader_name and self._size < 0), "When reader_name is provided, size should not be set"
+        if self._size < 0 and not reader_name:
+            self._auto_reset = False
+            self._fill_last_batch = False
+            self._last_batch_padded = False
+        self._pipes = pipelines
+        self._counter = 0
+
+        # Build all pipelines
+        for p in self._pipes:
+            with p._check_api_type_scope(types.PipelineAPIType.ITERATOR):
+                p.build()
+
+        self._reader_name = reader_name
+        if self._reader_name:
+            self._size_no_pad = self._pipes[0].epoch_size(self._reader_name, False)
+            assert np.all(np.equal([p.epoch_size(self._reader_name, False) for p in self._pipes], self._size_no_pad)), \
+                "All pipelines readers should have the same size, please check if they are reading the same data"
+
+            self._shards_num = self._pipes[0].shards_number(self._reader_name)
+            assert np.all(np.equal([p.shards_number(self._reader_name) for p in self._pipes], self._shards_num)), \
+                "All pipelines readers should have the same shard number set"
+
+            self._shards_id = np.array([p.shard_id(self._reader_name) for p in self._pipes], dtype=np.int)
+
+            assert np.all([p.is_pad_last_batch(self._reader_name) for p in self._pipes]) or \
+                   not np.any([p.is_pad_last_batch(self._reader_name) for p in self._pipes]), \
+                "All pipelines readers should have set padding in the same way"
```

Comment:
Maybe the error message should be something like
```
'Reader Operator "' + reader_name +'" should have the same `pad_last_batch` argument set in all the pipelines.'```

And similar for others.

Reply:
`epoch_size` doesn't follow that convention

Reply:
Done, I hope